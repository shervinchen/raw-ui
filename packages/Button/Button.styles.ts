import css from 'styled-jsx/css';
import { RawUITheme } from '../Theme/preset/preset.type';
import { useTheme } from '../Theme/theme-context';
import {
  ButtonSizes,
  ButtonTypes,
  ButtonVariants,
  ButtonProps,
  ButtonSizeStyles,
  ButtonBasicStyles,
  ButtonStyles,
  ButtonHoverStyles,
  ButtonActiveStyles,
} from './Button.types';

export const useButtonStyles = (props: ButtonProps): ButtonStyles => {
  const theme: RawUITheme = useTheme();
  const { size, type, variant, loading, disabled } = props;

  const sizes: {
    [key in ButtonSizes]: ButtonSizeStyles;
  } = {
    xs: {
      fontSize: '12px',
      height: '24px',
      horizontalPadding: '12px',
    },
    sm: {
      fontSize: '14px',
      height: '28px',
      horizontalPadding: '16px',
    },
    md: {
      fontSize: '14px',
      height: '32px',
      horizontalPadding: '16px',
    },
    lg: {
      fontSize: '14px',
      height: '36px',
      horizontalPadding: '16px',
    },
    xl: {
      fontSize: '16px',
      height: '40px',
      horizontalPadding: '20px',
    },
  };

  const styles: {
    [key in ButtonTypes]: {
      [key in ButtonVariants]: ButtonBasicStyles;
    };
  } = {
    default: {
      default: {
        backgroundColor: theme.palette.background,
        color: theme.palette.neutral['600'],
        borderColor: theme.palette.neutral['200'],
      },
      outline: {
        backgroundColor: theme.palette.background,
        color: theme.palette.neutral['600'],
        borderColor: theme.palette.neutral['200'],
      },
      ghost: {
        backgroundColor: 'transparent',
        color: theme.palette.neutral['600'],
        borderColor: 'transparent',
      },
      shadow: {
        backgroundColor: theme.palette.background,
        color: theme.palette.neutral['600'],
        borderColor: theme.palette.background,
        boxShadow: theme.tokens.shadow.sm,
      },
    },
    primary: {
      default: {
        backgroundColor: theme.palette.foreground,
        color: theme.palette.background,
        borderColor: theme.palette.foreground,
      },
      outline: {
        backgroundColor: 'transparent',
        color: theme.palette.foreground,
        borderColor: theme.palette.foreground,
      },
      ghost: {
        backgroundColor: 'transparent',
        color: theme.palette.foreground,
        borderColor: 'transparent',
      },
      shadow: {
        backgroundColor: theme.palette.foreground,
        color: theme.palette.background,
        borderColor: theme.palette.foreground,
        boxShadow: theme.tokens.shadow.sm,
      },
    },
    success: {
      default: {
        backgroundColor: theme.palette.blue['500'],
        color: theme.palette.background,
        borderColor: theme.palette.blue['500'],
      },
      outline: {
        backgroundColor: 'transparent',
        color: theme.palette.blue['500'],
        borderColor: theme.palette.blue['500'],
      },
      ghost: {
        backgroundColor: 'transparent',
        color: theme.palette.blue['500'],
        borderColor: 'transparent',
      },
      shadow: {
        backgroundColor: theme.palette.blue['500'],
        color: theme.palette.background,
        borderColor: theme.palette.blue['500'],
        boxShadow: theme.tokens.shadow.sm,
      },
    },
    warning: {
      default: {
        backgroundColor: theme.palette.yellow['500'],
        color: theme.palette.background,
        borderColor: theme.palette.yellow['500'],
      },
      outline: {
        backgroundColor: 'transparent',
        color: theme.palette.yellow['500'],
        borderColor: theme.palette.yellow['500'],
      },
      ghost: {
        backgroundColor: 'transparent',
        color: theme.palette.yellow['500'],
        borderColor: 'transparent',
      },
      shadow: {
        backgroundColor: theme.palette.yellow['500'],
        color: theme.palette.background,
        borderColor: theme.palette.yellow['500'],
        boxShadow: theme.tokens.shadow.sm,
      },
    },
    error: {
      default: {
        backgroundColor: theme.palette.red['500'],
        color: theme.palette.background,
        borderColor: theme.palette.red['500'],
      },
      outline: {
        backgroundColor: 'transparent',
        color: theme.palette.red['500'],
        borderColor: theme.palette.red['500'],
      },
      ghost: {
        backgroundColor: 'transparent',
        color: theme.palette.red['500'],
        borderColor: 'transparent',
      },
      shadow: {
        backgroundColor: theme.palette.red['500'],
        color: theme.palette.background,
        borderColor: theme.palette.red['500'],
        boxShadow: theme.tokens.shadow.sm,
      },
    },
  };

  const loadingStyles: {
    [key in ButtonTypes]: {
      [key in ButtonVariants]: ButtonBasicStyles;
    };
  } = {
    default: {
      default: {
        backgroundColor: theme.palette.background,
      },
      outline: {
        backgroundColor: theme.palette.background,
      },
      ghost: {
        backgroundColor: theme.palette.background,
      },
      shadow: {
        backgroundColor: theme.palette.background,
      },
    },
    primary: {
      default: {
        backgroundColor: theme.palette.foreground,
      },
      outline: {
        backgroundColor: theme.palette.background,
      },
      ghost: {
        backgroundColor: theme.palette.background,
      },
      shadow: {
        backgroundColor: theme.palette.foreground,
      },
    },
    success: {
      default: {
        backgroundColor: theme.palette.blue['500'],
      },
      outline: {
        backgroundColor: theme.palette.background,
      },
      ghost: {
        backgroundColor: theme.palette.background,
      },
      shadow: {
        backgroundColor: theme.palette.blue['500'],
      },
    },
    warning: {
      default: {
        backgroundColor: theme.palette.yellow['500'],
      },
      outline: {
        backgroundColor: theme.palette.background,
      },
      ghost: {
        backgroundColor: theme.palette.background,
      },
      shadow: {
        backgroundColor: theme.palette.yellow['500'],
      },
    },
    error: {
      default: {
        backgroundColor: theme.palette.red['500'],
      },
      outline: {
        backgroundColor: theme.palette.background,
      },
      ghost: {
        backgroundColor: theme.palette.background,
      },
      shadow: {
        backgroundColor: theme.palette.red['500'],
      },
    },
  };

  const disabledStyles: {
    [key in ButtonVariants]: ButtonBasicStyles;
  } = {
    default: {
      backgroundColor: theme.palette.neutral['200'],
      color: theme.palette.neutral['400'],
      borderColor: theme.palette.neutral['200'],
    },
    outline: {
      backgroundColor: theme.palette.neutral['50'],
      color: theme.palette.neutral['400'],
      borderColor: theme.palette.neutral['200'],
    },
    ghost: {
      color: theme.palette.neutral['400'],
    },
    shadow: {
      backgroundColor: theme.palette.neutral['200'],
      color: theme.palette.neutral['400'],
      borderColor: theme.palette.neutral['200'],
    },
  };

  const defaultStyles = {
    ...(styles?.[type || 'default']?.[variant || 'default'] ??
      styles['default']['default']),
    ...(sizes?.[size || 'md'] ?? sizes['md']),
  };

  return {
    ...defaultStyles,
    ...(loading
      ? {
          ...(loadingStyles?.[type || 'default']?.[variant || 'default'] ??
            loadingStyles['default']['default']),
          cursor: 'default',
        }
      : {}),
    ...(disabled
      ? {
          ...(type === 'default'
            ? disabledStyles['outline']
            : disabledStyles?.[variant || 'default'] ??
              disabledStyles['default']),
          cursor: 'not-allowed',
        }
      : {}),
  };
};

export const useButtonHoverStyles = (props: ButtonProps): ButtonHoverStyles => {
  const theme: RawUITheme = useTheme();
  const { type, variant, loading, disabled } = props;
  const styles: {
    [key in ButtonTypes]: {
      [key in ButtonVariants]: ButtonHoverStyles;
    };
  } = {
    default: {
      default: {
        hoverColor: theme.palette.foreground,
        hoverBorderColor: theme.palette.foreground,
      },
      outline: {
        hoverColor: theme.palette.foreground,
        hoverBorderColor: theme.palette.foreground,
      },
      ghost: {
        hoverBackgroundColor: theme.palette.neutral['200'],
        hoverColor: theme.palette.neutral['600'],
      },
      shadow: {
        hoverBoxShadow: theme.tokens.shadow.md,
        hoverTransform: 'translateY(-2px)',
      },
    },
    primary: {
      default: {
        hoverColor: theme.palette.foreground,
        hoverBackgroundColor: theme.palette.background,
      },
      outline: {
        hoverColor: theme.palette.background,
        hoverBackgroundColor: theme.palette.foreground,
      },
      ghost: {
        hoverBackgroundColor: theme.palette.neutral['200'],
      },
      shadow: {
        hoverBoxShadow: theme.tokens.shadow.md,
        hoverTransform: 'translateY(-2px)',
      },
    },
    success: {
      default: {
        hoverColor: theme.palette.blue['500'],
        hoverBackgroundColor: theme.palette.background,
      },
      outline: {
        hoverColor: theme.palette.background,
        hoverBackgroundColor: theme.palette.blue['500'],
      },
      ghost: {
        hoverBackgroundColor: theme.palette.blue['100'],
      },
      shadow: {
        hoverBoxShadow: theme.tokens.shadow.md,
        hoverTransform: 'translateY(-2px)',
      },
    },
    warning: {
      default: {
        hoverColor: theme.palette.yellow['500'],
        hoverBackgroundColor: theme.palette.background,
      },
      outline: {
        hoverColor: theme.palette.background,
        hoverBackgroundColor: theme.palette.yellow['500'],
      },
      ghost: {
        hoverBackgroundColor: theme.palette.yellow['100'],
      },
      shadow: {
        hoverBoxShadow: theme.tokens.shadow.md,
        hoverTransform: 'translateY(-2px)',
      },
    },
    error: {
      default: {
        hoverColor: theme.palette.red['500'],
        hoverBackgroundColor: theme.palette.background,
      },
      outline: {
        hoverColor: theme.palette.background,
        hoverBackgroundColor: theme.palette.red['500'],
      },
      ghost: {
        hoverBackgroundColor: theme.palette.red['100'],
      },
      shadow: {
        hoverBoxShadow: theme.tokens.shadow.md,
        hoverTransform: 'translateY(-2px)',
      },
    },
  };

  const defaultStyles =
    styles?.[type || 'default']?.[variant || 'default'] ??
    styles['default']['default'];

  return loading || disabled ? {} : defaultStyles;
};

export const useButtonActiveStyles = (
  props: ButtonProps
): ButtonActiveStyles => {
  const theme: RawUITheme = useTheme();
  const { type, variant, loading, disabled } = props;
  const styles: {
    [key in ButtonTypes]: {
      [key in ButtonVariants]: ButtonActiveStyles;
    };
  } = {
    default: {
      default: {
        activeBackgroundColor: theme.palette.neutral['200'],
      },
      outline: {
        activeBackgroundColor: theme.palette.neutral['200'],
      },
      ghost: {
        activeBackgroundColor: theme.palette.neutral['300'],
      },
      shadow: {
        activeBoxShadow: theme.tokens.shadow.sm,
        activeTransform: 'none',
      },
    },
    primary: {
      default: {
        activeBackgroundColor: theme.palette.neutral['200'],
      },
      outline: {
        activeBackgroundColor: theme.palette.neutral['800'],
      },
      ghost: {
        activeBackgroundColor: theme.palette.neutral['300'],
      },
      shadow: {
        activeBoxShadow: theme.tokens.shadow.sm,
        activeTransform: 'none',
      },
    },
    success: {
      default: {
        activeBackgroundColor: theme.palette.blue['200'],
      },
      outline: {
        activeBackgroundColor: theme.palette.blue['400'],
      },
      ghost: {
        activeBackgroundColor: theme.palette.blue['200'],
      },
      shadow: {
        activeBoxShadow: theme.tokens.shadow.sm,
        activeTransform: 'none',
      },
    },
    warning: {
      default: {
        activeBackgroundColor: theme.palette.yellow['200'],
      },
      outline: {
        activeBackgroundColor: theme.palette.yellow['400'],
      },
      ghost: {
        activeBackgroundColor: theme.palette.yellow['200'],
      },
      shadow: {
        activeBoxShadow: theme.tokens.shadow.sm,
        activeTransform: 'none',
      },
    },
    error: {
      default: {
        activeBackgroundColor: theme.palette.red['200'],
      },
      outline: {
        activeBackgroundColor: theme.palette.red['400'],
      },
      ghost: {
        activeBackgroundColor: theme.palette.red['200'],
      },
      shadow: {
        activeBoxShadow: theme.tokens.shadow.sm,
        activeTransform: 'none',
      },
    },
  };

  const defaultStyles =
    styles?.[type || 'default']?.[variant || 'default'] ??
    styles['default']['default'];

  return loading || disabled ? {} : defaultStyles;
};

export const useButtonCSS = (props: ButtonProps) => {
  const {
    fontSize,
    height,
    horizontalPadding,
    backgroundColor,
    borderColor,
    color,
    boxShadow = 'none',
    cursor = 'pointer',
  } = useButtonStyles(props);

  const {
    hoverBackgroundColor = backgroundColor,
    hoverBorderColor = borderColor,
    hoverColor = color,
    hoverBoxShadow = 'none',
    hoverTransform = 'none',
  } = useButtonHoverStyles(props);

  const {
    activeBackgroundColor = backgroundColor,
    activeBoxShadow = 'none',
    activeTransform = 'none',
  } = useButtonActiveStyles(props);

  return css.resolve`
    .raw-button:not(.raw-loading-button):not(.raw-disabled-button):hover,
    .raw-button:not(.raw-loading-button):not(.raw-disabled-button):focus {
      background-color: ${hoverBackgroundColor};
      border-color: ${hoverBorderColor};
      color: ${hoverColor};
      box-shadow: ${hoverBoxShadow};
      transform: ${hoverTransform};
    }

    .raw-button:not(.raw-loading-button):not(.raw-disabled-button):active {
      background-color: ${activeBackgroundColor};
      box-shadow: ${activeBoxShadow};
      transform: ${activeTransform};
    }

    .raw-button {
      box-sizing: border-box;
      position: relative;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex-shrink: 0;
      vertical-align: middle;
      border: 1px solid;
      border-radius: 6px;
      font-weight: 400;
      font-size: ${fontSize};
      line-height: 1;
      height: ${height};
      padding: 0 ${horizontalPadding};
      background-color: ${backgroundColor};
      border-color: ${borderColor};
      color: ${color};
      box-shadow: ${boxShadow};
      cursor: ${cursor};
      transition-property: border-color, background, color, transform,
        box-shadow;
      transition-duration: 0.2s;
      transition-timing-function: ease;
      text-transform: capitalize;
      white-space: nowrap;
      appearance: none;
      outline: none;
      user-select: none;
    }

    .raw-button.raw-childless-button {
      width: ${height};
      height: ${height};
      padding: 0;
    }

    .raw-button :global(.raw-button-content) {
      position: relative;
      z-index: 1;
    }
  `;
};
