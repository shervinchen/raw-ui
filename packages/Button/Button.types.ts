import { MouseEventHandler, ButtonHTMLAttributes, ReactNode } from 'react';

export type ButtonSizes = 'sm' | 'md' | 'lg';

export type ButtonTypes =
  | 'default'
  | 'primary'
  | 'success'
  | 'warning'
  | 'error';

export type ButtonVariants = 'default' | 'outline' | 'ghost' | 'shadow';

export interface BaseButtonProps {
  className?: string;
  size?: ButtonSizes;
  type?: ButtonTypes;
  variant?: ButtonVariants;
  htmlType?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
  loading?: boolean;
  disabled?: boolean;
  icon?: ReactNode;
  iconRight?: ReactNode;
  onClick?: MouseEventHandler<HTMLButtonElement>;
}

type NativeButtonProps = Omit<
  ButtonHTMLAttributes<HTMLButtonElement>,
  keyof BaseButtonProps
>;

export type ButtonProps = BaseButtonProps & NativeButtonProps;

export interface ButtonLoadingProps {
  color?: string;
  backgroundColor?: string;
}

export interface ButtonIconProps {
  isRight?: boolean;
  isSingle?: boolean;
  height?: string;
  horizontalPadding?: string;
}

export interface ButtonSizeStyles {
  fontSize?: string;
  lineHeight?: string;
  height?: string;
  horizontalPadding?: string;
}

export interface ButtonBasicStyles {
  backgroundColor?: string;
  borderColor?: string;
  color?: string;
  boxShadow?: string;
  cursor?: 'pointer' | 'default' | 'not-allowed';
}

export type ButtonStyles = ButtonSizeStyles & ButtonBasicStyles;

export interface ButtonHoverStyles {
  hoverBackgroundColor?: string;
  hoverBorderColor?: string;
  hoverColor?: string;
  hoverBoxShadow?: string;
  hoverTransform?: string;
}

export interface ButtonActiveStyles {
  activeBackgroundColor?: string;
  activeBoxShadow?: string;
  activeTransform?: string;
}
