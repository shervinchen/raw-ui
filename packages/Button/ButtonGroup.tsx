import React, { FC, PropsWithChildren, useMemo } from 'react';
import classNames from 'classnames';

import { ButtonGroupContext } from './button-group-context';
import { ButtonGroupProps, ButtonGroupConfig } from './ButtonGroup.types';
import { useButtonGroupCSS } from './ButtonGroup.styles';

const ButtonGroup: FC<PropsWithChildren<ButtonGroupProps>> = ({
  size = 'md',
  type = 'default',
  variant = 'default',
  disabled = false,
  children,
  className = '',
  vertical = false,
  ...resetProps
}) => {
  const initialConfig = useMemo<ButtonGroupConfig>(
    () => ({
      size,
      type,
      variant,
      disabled,
      isButtonGroup: true,
    }),
    []
  );

  const { className: resolveClassName, styles } = useButtonGroupCSS({
    type,
    size,
    variant,
    disabled,
  });

  const classes = classNames(
    'raw-button-group',
    vertical && 'vertical',
    !vertical && 'horizontal',
    className,
    resolveClassName
  );

  return (
    <ButtonGroupContext.Provider value={initialConfig}>
      <div className={classes} {...resetProps}>
        {children}
        {styles}
      </div>
    </ButtonGroupContext.Provider>
  );
};

export default ButtonGroup;
