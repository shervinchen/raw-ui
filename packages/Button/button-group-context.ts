import { createContext, useContext } from 'react';
import { ButtonGroupConfig } from './ButtonGroup.types';

const defaultContext: ButtonGroupConfig = {
  size: 'md',
  type: 'default',
  variant: 'default',
  disabled: false,
  isButtonGroup: false,
};

export const ButtonGroupContext =
  createContext<ButtonGroupConfig>(defaultContext);

export const useButtonGroupContext = (): ButtonGroupConfig =>
  useContext<ButtonGroupConfig>(ButtonGroupContext);
