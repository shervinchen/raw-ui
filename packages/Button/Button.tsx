import React, {
  forwardRef,
  Children,
  MouseEvent,
  useRef,
  useImperativeHandle,
  ComponentPropsWithRef,
} from 'react';
import classNames from 'classnames';

import { useButtonStyles, useButtonCSS } from './Button.styles';
import { ButtonProps } from './Button.types';
import ButtonLoading from './ButtonLoading';
import ButtonIcon from './ButtonIcon';
import { useButtonGroupContext } from './button-group-context';
import { ButtonGroupConfig } from './ButtonGroup.types';

const mergeButtonGroupProps = (
  buttonProps: ButtonProps,
  config: ButtonGroupConfig
): ButtonProps => {
  if (!config.isButtonGroup) return buttonProps;
  return {
    ...buttonProps,
    size: config.size,
    type: config.type,
    variant: config.variant,
    disabled: config.disabled,
  };
};

const Button = forwardRef(
  (buttonProps: ButtonProps, ref: ComponentPropsWithRef<'button'>['ref']) => {
    const buttonGroupConfig = useButtonGroupContext();
    const {
      className = '',
      size = 'md',
      type = 'default',
      variant = 'default',
      htmlType = 'button',
      loading = false,
      disabled = false,
      icon,
      iconRight,
      onClick,
      children,
      ...restProps
    } = mergeButtonGroupProps(buttonProps, buttonGroupConfig);
    const buttonRef = useRef<HTMLButtonElement>(null);

    const { horizontalPadding, height, color, backgroundColor } =
      useButtonStyles({
        type,
        size,
        variant,
        loading,
        disabled,
      });

    const { className: resolveClassName, styles } = useButtonCSS({
      type,
      size,
      variant,
      loading,
      disabled,
    });

    const isChildLess = Children.count(children) === 0;
    const isRight = Boolean(iconRight);

    const classes = classNames(
      'raw-button',
      loading && 'raw-loading-button',
      disabled && 'raw-disabled-button',
      isChildLess && 'raw-childless-button',
      className,
      resolveClassName
    );

    const handleClick = (event: MouseEvent<HTMLButtonElement>) => {
      if (loading || disabled) return;
      onClick?.(event);
    };

    useImperativeHandle(ref, () => buttonRef.current);

    return (
      <button
        ref={buttonRef}
        type={htmlType}
        className={classes}
        onClick={handleClick}
        disabled={disabled}
        aria-disabled={disabled}
        {...restProps}
      >
        {loading && (
          <ButtonLoading color={color} backgroundColor={backgroundColor} />
        )}
        {icon && (
          <ButtonIcon
            isSingle={isChildLess}
            height={height}
            horizontalPadding={horizontalPadding}
          >
            {icon}
          </ButtonIcon>
        )}
        {!isChildLess && (
          <span className="raw-button-content" data-testid="buttonContent">
            {children}
          </span>
        )}
        {iconRight && (
          <ButtonIcon
            isSingle={isChildLess}
            isRight={isRight && !isChildLess}
            height={height}
            horizontalPadding={horizontalPadding}
          >
            {iconRight}
          </ButtonIcon>
        )}
        {styles}
      </button>
    );
  }
);

Button.displayName = 'RawButton';

export default Button;
