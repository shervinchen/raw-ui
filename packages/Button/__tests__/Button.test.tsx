import React from 'react';
import { render, screen, renderHook } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Search } from 'react-feather';
import Button from '..';
import { ButtonSizes, ButtonTypes, ButtonVariants } from '../Button.types';
import {
  useButtonActiveStyles,
  useButtonHoverStyles,
  useButtonStyles,
} from '../Button.styles';

const typeColorMap = {
  primary: 'rgb(0, 0, 0)',
  success: 'rgb(0, 112, 243)',
  warning: 'rgb(245, 166, 35)',
  error: 'rgb(238, 0, 0)',
};

const sizeHeightMap = {
  sm: '34px',
  md: '40px',
  lg: '46px',
};

const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

describe('Button', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(<Button>Default</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should forward ref', () => {
    const ref = React.createRef<HTMLButtonElement>();
    render(<Button ref={ref}>Text</Button>);
    const button = screen.getByRole('button', { name: /Text/i });
    expect(button).toEqual(ref.current);
  });

  test('should support custom class name', () => {
    render(<Button className="custom-button">Text</Button>);
    const button = screen.getByRole('button', { name: /Text/i });
    expect(button).toHaveClass('custom-button');
  });

  test('should trigger event when clicked', async () => {
    const clickHandler = jest.fn();
    render(<Button onClick={clickHandler}>Text</Button>);
    const button = screen.getByRole('button', { name: /Text/i });
    await user.click(button);
    expect(clickHandler).toHaveBeenCalledTimes(1);
  });

  ['primary', 'success', 'warning', 'error'].forEach(
    (item: Exclude<ButtonTypes, 'default'>) => {
      test(`should render ${item} type`, () => {
        render(<Button type={item}>Text</Button>);
        const button = screen.getByRole('button', { name: /Text/i });
        expect(button).toHaveStyle(`background-color: ${typeColorMap[item]}`);
      });
    }
  );

  ['sm', 'md', 'lg'].forEach((item: ButtonSizes) => {
    test(`should render ${item} size`, () => {
      render(<Button size={item}>Text</Button>);
      const button = screen.getByRole('button', { name: /Text/i });
      expect(button).toHaveStyle(`height: ${sizeHeightMap[item]}`);
    });
  });

  test('should render outline variant', () => {
    render(
      <Button type="primary" variant="outline">
        Text
      </Button>
    );
    const button = screen.getByRole('button', { name: /Text/i });
    expect(button).toHaveStyle({
      color: 'rgb(0, 0, 0)',
      borderColor: '#000000',
    });
  });

  test('should render ghost variant', () => {
    render(
      <Button type="primary" variant="ghost">
        Text
      </Button>
    );
    const button = screen.getByRole('button', { name: /Text/i });
    expect(button).toHaveStyle({
      backgroundColor: 'transparent',
      borderColor: 'transparent',
    });
  });

  test('should render shadow variant', () => {
    render(
      <Button type="primary" variant="shadow">
        Text
      </Button>
    );
    const button = screen.getByRole('button', { name: /Text/i });
    expect(button).toHaveStyle('boxShadow: 0 5px 10px rgba(0, 0, 0, 0.12)');
  });

  test('should get default style when type is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useButtonStyles({
        type: 'unknown' as ButtonTypes,
        size: 'md',
        variant: 'default',
        loading: false,
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useButtonStyles({
        type: undefined as unknown as ButtonTypes,
        size: 'md',
        variant: 'default',
        loading: false,
        disabled: false,
      })
    );
    expect(result1.current.backgroundColor).toBe('#ffffff');
    expect(result2.current.backgroundColor).toBe('#ffffff');
  });

  test('should get default loading style when type is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useButtonStyles({
        type: 'unknown' as ButtonTypes,
        size: 'md',
        variant: 'default',
        loading: true,
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useButtonStyles({
        type: undefined as unknown as ButtonTypes,
        size: 'md',
        variant: 'default',
        loading: true,
        disabled: false,
      })
    );
    expect(result1.current.backgroundColor).toBe('#ffffff');
    expect(result2.current.backgroundColor).toBe('#ffffff');
  });

  test('should get default style when variant is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useButtonStyles({
        type: 'default',
        size: 'md',
        variant: 'unknown' as ButtonVariants,
        loading: false,
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useButtonStyles({
        type: 'default',
        size: 'md',
        variant: undefined as unknown as ButtonVariants,
        loading: false,
        disabled: false,
      })
    );
    expect(result1.current.backgroundColor).toBe('#ffffff');
    expect(result2.current.backgroundColor).toBe('#ffffff');
  });

  test('should get default loading style when variant is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useButtonStyles({
        type: 'default',
        size: 'md',
        variant: 'unknown' as ButtonVariants,
        loading: true,
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useButtonStyles({
        type: 'default',
        size: 'md',
        variant: undefined as unknown as ButtonVariants,
        loading: true,
        disabled: false,
      })
    );
    expect(result1.current.backgroundColor).toBe('#ffffff');
    expect(result2.current.backgroundColor).toBe('#ffffff');
  });

  test('should get default disabled style when type is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useButtonStyles({
        type: 'unknown' as ButtonTypes,
        size: 'md',
        variant: 'default',
        loading: false,
        disabled: true,
      })
    );
    const { result: result2 } = renderHook(() =>
      useButtonStyles({
        type: undefined as unknown as ButtonTypes,
        size: 'md',
        variant: 'default',
        loading: false,
        disabled: true,
      })
    );
    expect(result1.current.backgroundColor).toBe('#eaeaea');
    expect(result2.current.backgroundColor).toBe('#eaeaea');
  });

  test('should get default disabled style when variant is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useButtonStyles({
        type: undefined as unknown as ButtonTypes,
        size: 'md',
        variant: 'unknown' as ButtonVariants,
        loading: false,
        disabled: true,
      })
    );
    const { result: result2 } = renderHook(() =>
      useButtonStyles({
        type: undefined as unknown as ButtonTypes,
        size: 'md',
        variant: undefined as unknown as ButtonVariants,
        loading: false,
        disabled: true,
      })
    );
    expect(result1.current.backgroundColor).toBe('#eaeaea');
    expect(result2.current.backgroundColor).toBe('#eaeaea');
  });

  test('should get default hover style when type is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useButtonHoverStyles({
        type: 'unknown' as ButtonTypes,
        size: 'md',
        variant: 'default',
        loading: false,
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useButtonHoverStyles({
        type: undefined as unknown as ButtonTypes,
        size: 'md',
        variant: 'default',
        loading: false,
        disabled: false,
      })
    );
    expect(result1.current.hoverColor).toBe('#000000');
    expect(result2.current.hoverColor).toBe('#000000');
  });

  test('should get default hover style when variant is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useButtonHoverStyles({
        type: 'default',
        size: 'md',
        variant: 'unknown' as ButtonVariants,
        loading: false,
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useButtonHoverStyles({
        type: 'default',
        size: 'md',
        variant: undefined as unknown as ButtonVariants,
        loading: false,
        disabled: false,
      })
    );
    expect(result1.current.hoverColor).toBe('#000000');
    expect(result2.current.hoverColor).toBe('#000000');
  });

  test('should get default active style when type is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useButtonActiveStyles({
        type: 'unknown' as ButtonTypes,
        size: 'md',
        variant: 'default',
        loading: false,
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useButtonActiveStyles({
        type: undefined as unknown as ButtonTypes,
        size: 'md',
        variant: 'default',
        loading: false,
        disabled: false,
      })
    );
    expect(result1.current.activeBackgroundColor).toBe('#eaeaea');
    expect(result2.current.activeBackgroundColor).toBe('#eaeaea');
  });

  test('should get default active style when variant is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useButtonActiveStyles({
        type: 'default',
        size: 'md',
        variant: 'unknown' as ButtonVariants,
        loading: false,
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useButtonActiveStyles({
        type: 'default',
        size: 'md',
        variant: undefined as unknown as ButtonVariants,
        loading: false,
        disabled: false,
      })
    );
    expect(result1.current.activeBackgroundColor).toBe('#eaeaea');
    expect(result2.current.activeBackgroundColor).toBe('#eaeaea');
  });

  test('should get md size when size is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useButtonStyles({
        type: 'default',
        size: 'unknown' as ButtonSizes,
        variant: 'default',
        loading: false,
        disabled: false,
      })
    );
    const { result: result2 } = renderHook(() =>
      useButtonStyles({
        type: 'default',
        size: undefined as unknown as ButtonSizes,
        variant: 'default',
        loading: false,
        disabled: false,
      })
    );
    expect(result1.current.height).toBe('40px');
    expect(result2.current.height).toBe('40px');
  });

  test('should support loading', async () => {
    const clickHandler = jest.fn();
    render(
      <Button loading onClick={clickHandler}>
        Text
      </Button>
    );
    const button = screen.getByRole('button', { name: /Text/i });
    expect(button).toHaveClass('raw-loading-button');
    await user.click(button as Element);
    expect(clickHandler).toHaveBeenCalledTimes(0);
  });

  test('should support disabled', () => {
    render(<Button disabled>Text</Button>);
    const button = screen.getByRole('button', { name: /Text/i });
    expect(button).toHaveClass('raw-disabled-button');
    expect(button).toBeDisabled();
  });

  test('should support icon without text', () => {
    render(<Button icon={<Search />} />);
    expect(screen.getByTestId('buttonIcon')).toBeInTheDocument();
    expect(screen.queryByTestId('buttonContent')).not.toBeInTheDocument();
    expect(screen.getByTestId('buttonIcon')).toHaveClass('button-icon-single');
  });

  test('should support icon with text', () => {
    render(<Button icon={<Search />}>Search</Button>);
    expect(screen.getByTestId('buttonIcon')).toBeInTheDocument();
    expect(screen.queryByTestId('buttonContent')).toBeInTheDocument();
    expect(screen.getByTestId('buttonIcon')).not.toHaveClass(
      'button-icon-single'
    );
  });

  test('should support right icon', () => {
    render(<Button iconRight={<Search />}>Search</Button>);
    expect(screen.getByTestId('buttonIcon')).toBeInTheDocument();
    expect(screen.getByTestId('buttonIcon')).not.toHaveClass(
      'button-icon-single'
    );
    expect(screen.getByTestId('buttonIcon')).toHaveClass('button-icon-right');
  });
});
