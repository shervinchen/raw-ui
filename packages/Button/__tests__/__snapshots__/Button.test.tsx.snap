// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="raw-button raw-md-button jsx-2168746361"
    type="button"
  >
    <span
      class="raw-button-content"
    >
      Default
    </span>
  </button>
</DocumentFragment>
`;
